cpp/tests/experimental/spatial/linestring_intersection_with_duplicates_test.cu

```diff
@@ -0,0 +1,812 @@
+
+  using index_t = std::size_t;
+
+  auto multilinestrings1 = make_multilinestring_array(
+    {0, 1}, {0, 3}, std::initializer_list<P>{P{-1, 0}, P{0, 0}, P{1, 1}});
+
+  auto multilinestrings2 =
+    make_multilinestring_array({0, 1}, {0, 3}, std::initializer_list<P>{P{0, 2}, P{0, 1}, P{1, 0}});
+
+  auto [points, segments] = pairwise_linestring_intersection_with_duplicates<index_t, T>(
+    multilinestrings1.range(), multilinestrings2.range(), this->mr(), this->stream());
+
+  auto expected_points_offsets = make_device_vector<index_t>({0, 1});
+  auto expected_points_coords  = make_device_vector<P>({P{0.5, 0.5}});
+
+  auto expected_segments_offsets = make_device_vector<index_t>({0, 0});
+  auto expected_segments_coords  = make_device_vector<segment<T>>({});
+
+  auto expected_point_lhs_linestring_ids = make_device_vector<index_t>({0});
+  auto expected_point_lhs_segment_ids    = make_device_vector<index_t>({1});
+
+  auto expected_point_rhs_linestring_ids = make_device_vector<index_t>({0});
+  auto expected_point_rhs_segment_ids    = make_device_vector<index_t>({1});
+
+  auto expected_segment_lhs_linestring_ids = make_device_vector<index_t>({});
+  auto expected_segment_lhs_segment_ids    = make_device_vector<index_t>({});
+
+  auto expected_segment_rhs_linestring_ids = make_device_vector<index_t>({});
+  auto expected_segment_rhs_segment_ids    = make_device_vector<index_t>({});
+
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_points_offsets, *std::move(points.offsets));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_points_coords, *std::move(points.geoms));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_segments_offsets, *std::move(segments.offsets));
+  expect_segment_equivalent(expected_segments_coords, *std::move(segments.geoms));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_point_lhs_linestring_ids,
+                                      *std::move(points.lhs_linestring_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_point_lhs_segment_ids,
+                                      *std::move(points.lhs_segment_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_point_rhs_linestring_ids,
+                                      *std::move(points.rhs_linestring_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_point_rhs_segment_ids,
+                                      *std::move(points.rhs_segment_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_segment_lhs_linestring_ids,
+                                      *std::move(segments.lhs_linestring_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_segment_lhs_segment_ids,
+                                      *std::move(segments.lhs_segment_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_segment_rhs_linestring_ids,
+                                      *std::move(segments.rhs_linestring_ids));
+  CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(expected_segment_rhs_segment_ids,
+                                      *std::move(segments.rhs_segment_ids));
```

Comment:
A lot of duplication in these tests. Perhaps a test function that parameterizes the things that change between tests and then call that function from each test.

Reply:
OK, but you still get line numbers, and combined with the failing test name you can easily jump to the test case that failed and which vector mismatched within it. 

But even if you duplicate the EXPECT macro calls, there seems to be a lot of unnecessary duplication in the declaration of all the test and expected vectors.

Reply:
That is because the tests are... duplicated! Yeah I agree the duplication is a bit hard to read. I'll work on some refactor on the next round of update.